### PR TITLE
Pre-populate workers with existing router on startup

### PR DESCRIPTION
**Pull request by rodis**
DHC-1378

Create a new thread that runs at startup and fetch
existing routers from quantum pushing healt_check message
in the scheduler queue for every router.

Signed-off-by: Rosario Di Somma rosario.disomma@dreamhost.com

### CODE REVIEW

**Comment by rodis** [comment]
I need some more tests, but f you wanna give a first look, the implementation should be ready.

File: akanda/rug/populate.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    nap_time = 1
+    max_sleep = 15
+    keep_going = True
+
+    quantum_client = quantum.Quantum(cfg.CONF)
+
+    while keep_going:
+        try:
+            quantum_routers = quantum_client.get_routers()
+            keep_going = False
```
Comment (dhellmann) [inline akanda/rug/populate.py]: This could just be a break statement.

File: akanda/rug/populate.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        except quantum.client.exceptions.QuantumClientException as err:
```
Comment (dhellmann) [inline akanda/rug/populate.py]: In the old implementation we only ignored QuantumClientException. Do we want to ignore any other errors here in the new version?

File: akanda/rug/test/unit/test_populate.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+        quantum_client = mock.Mock()
+        returned_value = [quantum.client.exceptions.QuantumClientException, []]
+        quantum_client.get_routers.side_effect = returned_value
+
+        mocked_quantum.side_effect = lambda *args, **kwargs: quantum_client
+
+        sched = mock.Mock()
+        populate._pre_populate_workers(sched)
+        self.assertEqual(
```
Comment (dhellmann) [inline akanda/rug/test/unit/test_populate.py:22]: Nice use of mocks here in these tests. I need to spend more time reading the docs for that module to see how to do stuff like this.

File: akanda/rug/test/unit/test_populate.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        sched = mock.Mock()
+        populate._pre_populate_workers(sched)
+        expected = [
+            mock.call.warning(u'Could not fetch routers from quantum: '
+                              'An unknown exception occurred.'),
+            mock.call.warning('sleeping 1 seconds before retrying'),
+            mock.call.debug('Start pre-populating the workers '
+                            'with %d fetched routers', 0)
+        ]
+        self.assertEqual(log.mock_calls, expected)
```
Comment (ryanpetrello) [inline akanda/rug/test/unit/test_populate.py:76]: Could we add a test that _doesn't_ return a `QuantumClientException` and tests that `scheduler.handle_message` is called with the proper arguments for each router?

Comment (rodis) [inline akanda/rug/test/unit/test_populate.py:76]: Yep, I'm adding other tests

File: akanda/rug/populate.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    nap_time = 1
+    max_sleep = 15
+
+    quantum_client = quantum.Quantum(cfg.CONF)
+
+    while True:
+        try:
+            quantum_routers = quantum_client.get_routers()
+            break
+        except Exception as err:
```
Comment (dhellmann) [inline akanda/rug/populate.py:38]: Now that I think of it, if this gets an authentication error it's never going to work, so it should probably treat that as a special case and return so the thread exits.

**Pull request merged**